asic/src/tofino_asic/ports.rs

```diff
@@ -412,6 +428,118 @@ pub fn set_prbs(
+            },
+        },
+    }
+}
+
+// Helper struct to return the lanes in the `bf_tof2_ber_t` array fields that
+// are expected to be non-zero. This is based entirely on the speed of the link.
+// See
+// https://github.com/oxidecomputer/tofino-sde/blob/0e4cfe10c92db223a953923d77f09a299b82157b/pkgsrc/bf-drivers/src/bf_pm/pm.c#L2120-L2140
+// for the logic.
+struct BerLanes {
+    start: usize,
+    end: usize,
+}
+
+impl BerLanes {
+    const fn all() -> Self {
+        Self {
+            start: 0,
+            end: MAX_N_LANES - 1,
+        }
+    }
+
+    // Return the lanes used for BER counters for the provided speed.
+    //
+    // Use all the counters if we can't reliably determine which ones to read.
+    fn new(hdl: PortHdl, speed: PortSpeed) -> Self {
+        // This magic logic appears in `pm_port_tof2_ber_get()`, where it
+        // appears to be used to determine the right counters to read from the
+        // serdes firmware.
+        const MASK: u64 = 0x07;
+        const FACTOR: u64 = 2;
+        let Ok(start) = usize::try_from((u64::from(hdl) & MASK) * FACTOR)
+        else {
+            return Self::all();
+        };
+        match speed {
+            PortSpeed::Speed0G
+            | PortSpeed::Speed1G
+            | PortSpeed::Speed10G
+            | PortSpeed::Speed25G
+            | PortSpeed::Speed40G => Self::all(),
+            PortSpeed::Speed50G => Self {
+                start,
+                end: start + 2,
+            },
+            PortSpeed::Speed100G => Self {
+                start,
+                end: start + 4,
+            },
+            PortSpeed::Speed200G => Self {
+                start,
+                end: start + 8,
+            },
+            PortSpeed::Speed400G => Self::all(),
+        }
+    }
+}
+
+/// Return the bit-error for a port.
+pub fn get_ber(hdl: &Handle, port_hdl: PortHdl) -> AsicResult<Ber> {
+    // Extract the BER and symbol error counts themselves.
+    let dev = hdl.dev_id;
+    let mut fp = FrontPortHandle::from_port_hdl(hdl, port_hdl)?;
+    let mut ber = empty_ber();
+    let ber = match unsafe {
+        bf_pm_port_ber_get(dev, fp.ptr(), &mut ber as *mut _)
+    } {
+        BF_SUCCESS => {
+            // Safety: We're using the `empty_ber()` function above to
+            // initialize this, so we know this field exists and is the active
+            // union field. It's also the only one.
+            unsafe { ber.ber.tof2 }
+        }
+        e => {
+            slog::error!(hdl.log, "failed to get port BER"; "error" => %e);
+            return Err(sde_error("getting port BER", e));
+        }
+    };
+
+    // Fetch the port speed, which is needed to determine which counters from
+    // the above call to `bf_pm_port_ber_get` are valid for this port.
+    let bf_speed = bf_port_speed_e_BF_SPEED_NONE;
+    let dev_port = fp.get_dev_port()?.into();
+    let res =
+        unsafe { bf_port_speed_get(dev, dev_port, &bf_speed as *const _ as _) };
```

Comment:
Don't we already know the port speed?

Reply:
Not really, at this layer of the code anyway. We could pass it down in the interface to the new BER trait method, it just seemed less helpful of an interface. I don't have a strong preference, we can do it that way too.

Reply:
Obviated by [5ce22d7](https://github.com/oxidecomputer/dendrite/pull/20/commits/5ce22d7c672cf3fcffce5bda96cb4f2871d766f6)